File: apps/web/lib/api/codex.ts
/**
 * Typed endpoint functions + small helpers for the Codex resource.
 *
 * Codex entries are PROJECT-scoped (`apps/api/app/api/v1/codex.py`, prefix
 * `/projects/{project_id}/codex` under `/api/v1`):
 *   GET    /projects/{pid}/codex            → CodexEntryRead[]
 *   GET    /projects/{pid}/codex/{id}       → CodexEntryRead
 *   POST   /projects/{pid}/codex            → CodexEntryRead (201)
 *   PATCH  /projects/{pid}/codex/{id}       → CodexEntryRead
 *   DELETE /projects/{pid}/codex/{id}       → 204
 *
 * BACKEND CONTRACT (apps/api/app/schemas/codex_entry.py): since P1.4 a Codex card
 * carries DEDICATED `aliases: string[]` (recognition names) + `role: string|null`
 * (the single story role) columns, mirroring `Character`. `content` is the
 * DESCRIPTION, `ai_visible` is the spoiler-protection AI gate, and `tags` is a
 * list of plain user labels.
 *
 * HISTORY: M6 had no `aliases` / `role` columns, so the UI folded those (plus a
 * "name-tracking off" flag) into `tags` via a `__woa:` namespaced codec. P1.4
 * added the real columns and REMOVED that codec entirely — aliases + role now
 * come from / go to their own fields, and `tags` is purely user labels. The M6
 * "track by name" toggle had no backend gate (the real AI gate is `ai_visible`,
 * a separate column), so it is now UI-only local state in the Tracking tab — no
 * residual `__woa:` tag is written. See `components/codex/codex-detail.tsx`.
 */
import { apiFetch } from "./client";
import {
  codexEntryListSchema,
  codexEntryReadSchema,
  codexEntryCreateSchema,
  codexEntryUpdateSchema,
  type CodexEntryCreate,
  type CodexEntryRead,
  type CodexEntryUpdate,
} from "./types";

/* ---------------------------------------------------------------------------
 * Entry types (mirrors CLAUDE.md + the prototype's type picker).
 * ------------------------------------------------------------------------- */

/** The codex entry types the New-Codex modal offers (slug = `entry_type`). */
export const CODEX_ENTRY_TYPES = [
  "character",
  "location",
  "object",
  "organization",
  "lore",
  "rule",
] as const;

export type CodexEntryType = (typeof CODEX_ENTRY_TYPES)[number];

/** Narrow an arbitrary `entry_type` string to a known type, or null. */
export function asCodexEntryType(value: string): CodexEntryType | null {
  return (CODEX_ENTRY_TYPES as readonly string[]).includes(value)
    ? (value as CodexEntryType)
    : null;
}

/* ---------------------------------------------------------------------------
 * Alias input parsing — comma-separated free text → a clean alias list.
 * ------------------------------------------------------------------------- */

/** Parse a comma-separated alias input into a deduped, trimmed list. */
export function parseAliasInput(raw: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const part of raw.split(",")) {
    const value = part.trim();
    if (value && !seen.has(value.toLowerCase())) {
      seen.add(value.toLowerCase());
      out.push(value);
    }
  }
  return out;
}

/* ---------------------------------------------------------------------------
 * Manuscript mention scan — the single source of truth for "N megemlítés".
 * Shared by the detail header AND the sidebar so the count is the same real
 * scene-content scan everywhere (NOT the alias count). MVP corpus is small, so
 * one pass over the already-cached book tree is cheap.
 * ------------------------------------------------------------------------- */

/** The recognition needles for an entry: its title + decoded aliases. */
export function mentionNeedles(
  title: string,
  aliases: readonly string[],
): string[] {
  return [title, ...aliases]
    .map((n) => n.trim())
    .filter((n) => n.length > 0);
}

/** Case-insensitive containment of any needle in a scene's content. */
export function contentMentions(
  content: string | null | undefined,
  needles: readonly string[],
): boolean {
  if (!content) return false;
  const haystack = content.toLowerCase();
  return needles.some((n) => haystack.includes(n.toLowerCase()));
}

/** The minimal scene shape the mention scan reads (chapter+scene tree). */
export interface MentionScene {
  content: string | null;
}
export interface MentionChapter {
  scenes: readonly MentionScene[];
}

/** Count scenes (across the book tree) that mention any of `needles`. */
export function countMentions(
  chapters: readonly MentionChapter[],
  needles: readonly string[],
): number {
  if (needles.length === 0) return 0;
  let count = 0;
  for (const chapter of chapters) {
    for (const scene of chapter.scenes) {
      if (contentMentions(scene.content, needles)) count += 1;
    }
  }
  return count;
}

/* ---------------------------------------------------------------------------
 * Endpoint functions (each validates the response with Zod — drift throws).
 * ------------------------------------------------------------------------- */

/** List the codex entries belonging to a project. */
export async function listCodexEntries(
  projectId: string,
): Promise<CodexEntryRead[]> {
  const data = await apiFetch<unknown>(`/projects/${projectId}/codex`);
  return codexEntryListSchema.parse(data);
}

/** Fetch a single codex entry by id. */
export async function getCodexEntry(
  projectId: string,
  entryId: string,
): Promise<CodexEntryRead> {
  const data = await apiFetch<unknown>(
    `/projects/${projectId}/codex/${entryId}`,
  );
  return codexEntryReadSchema.parse(data);
}

/** Create a codex entry under a project. The payload is validated first. */
export async function createCodexEntry(
  projectId: string,
  input: CodexEntryCreate,
): Promise<CodexEntryRead> {
  const body = codexEntryCreateSchema.parse(input);
  const data = await apiFetch<unknown>(`/projects/${projectId}/codex`, {
    method: "POST",
    body,
  });
  return codexEntryReadSchema.parse(data);
}

/** Patch a codex entry (name / type / description / ai_visible / tags). */
export async function updateCodexEntry(
  projectId: string,
  entryId: string,
  patch: CodexEntryUpdate,
): Promise<CodexEntryRead> {
  const body = codexEntryUpdateSchema.parse(patch);
  const data = await apiFetch<unknown>(
    `/projects/${projectId}/codex/${entryId}`,
    { method: "PATCH", body },
  );
  return codexEntryReadSchema.parse(data);
}

/** Delete a codex entry (the backend answers 204; apiFetch returns null). */
export async function deleteCodexEntry(
  projectId: string,
  entryId: string,
): Promise<void> {
  await apiFetch<unknown>(`/projects/${projectId}/codex/${entryId}`, {
    method: "DELETE",
  });
}

File: apps/api/app/api/v1/codex.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user, get_db
from app.schemas.codex_entry import CodexEntryCreate, CodexEntryRead, CodexEntryUpdate
from app.services.crud_codex_entry import (
    create_codex_entry,
    delete_codex_entry,
    get_codex_entry,
    list_codex_entries,
    update_codex_entry,
)
from app.services.crud_project import get_project
from app.services.crud_series import SeriesScopeError

router = APIRouter(prefix="/projects/{project_id}/codex", tags=["codex"])


async def _get_project_or_404(project_id: uuid.UUID, db: AsyncSession):
    project = await get_project(db, project_id)
    if project is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")
    return project


@router.post("", response_model=CodexEntryRead, status_code=status.HTTP_201_CREATED)
async def create(
    project_id: uuid.UUID,
    data: CodexEntryCreate,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexEntryRead:
    await _get_project_or_404(project_id, db)
    try:
        return await create_codex_entry(db, project_id, data)
    except SeriesScopeError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)
        ) from exc


@router.get("", response_model=list[CodexEntryRead])
async def list_all(
    project_id: uuid.UUID,
    series_id: uuid.UUID | None = Query(
        None,
        description=(
            "Optional series scope filter. When given, returns project-global "
            "entries (series_id IS NULL) PLUS entries scoped to this series, and "
            "EXCLUDES entries scoped to any other series. Omit to return all "
            "entries in the project."
        ),
    ),
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> list[CodexEntryRead]:
    await _get_project_or_404(project_id, db)
    return await list_codex_entries(db, project_id, series_id)


@router.get("/{entry_id}", response_model=CodexEntryRead)
async def get_one(
    project_id: uuid.UUID,
    entry_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexEntryRead:
    await _get_project_or_404(project_id, db)
    entry = await get_codex_entry(db, project_id, entry_id)
    if entry is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Codex entry not found")
    return entry


@router.patch("/{entry_id}", response_model=CodexEntryRead)
async def update(
    project_id: uuid.UUID,
    entry_id: uuid.UUID,
    data: CodexEntryUpdate,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> CodexEntryRead:
    await _get_project_or_404(project_id, db)
    entry = await get_codex_entry(db, project_id, entry_id)
    if entry is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Codex entry not found")
    try:
        return await update_codex_entry(db, entry, data)
    except SeriesScopeError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)
        ) from exc


@router.delete("/{entry_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(
    project_id: uuid.UUID,
    entry_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    _: str = Depends(get_current_user),
) -> None:
    await _get_project_or_404(project_id, db)
    entry = await get_codex_entry(db, project_id, entry_id)
    if entry is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Codex entry not found")
    await delete_codex_entry(db, entry)
